src/dodal/devices/i09_1_shared/hard_undulator_functions.py

```diff
@@ -0,0 +1,111 @@
+import numpy as np
+
+from dodal.devices.util.lookup_tables import energy_distance_table
+from dodal.log import LOGGER
+
+LUT_COMMENTS = ["#"]
+HU_SKIP_ROWS = 3
+
+# Physics constants
+ELECTRON_REST_ENERGY_MEV = 0.510999
+
+# Columns in the lookup table
+RING_ENERGY_COLUMN = 1
+MAGNET_FIELD_COLUMN = 2
+MIN_ENERGY_COLUMN = 3
+MAX_ENERGY_COLUMN = 4
+GAP_OFFSET_COLUMN = 7
+
+
+async def get_hu_lut_as_dict(lut_path: str) -> dict:
+    lut_dict: dict = {}
+    _lookup_table: np.ndarray = await energy_distance_table(
+        lut_path,
+        comments=LUT_COMMENTS,
+        skiprows=HU_SKIP_ROWS,
+    )
+    for i in range(_lookup_table.shape[0]):
+        lut_dict[_lookup_table[i][0]] = _lookup_table[i]
+        LOGGER.debug(f"Loaded lookup table:\n {lut_dict}")
+    return lut_dict
+
+
+def calculate_gap_hu(
+    photon_energy_kev: float,
+    look_up_table: dict[int, "np.ndarray"],
+    order: int = 1,
+    gap_offset: float = 0.0,
+    undulator_period_mm: int = 27,
+) -> float:
+    """
+    Calculate the undulator gap required to produce a given energy at a given harmonic order.
```

Comment:
This is really great, thank you! Really good example of how we're improving things as we move across from GDA